src/libstd/time/duration.rs

```diff
@@ -168,15 +173,114 @@ impl Div<u32> for Duration {
+                    // overwrite it, and then restore it.
+                    let n = n as f64 / NANOS_PER_SEC as f64;
+                    try!(cursor.seek(SeekFrom::Current(-1)));
+                    let saved_pos = cursor.position() as usize;
+                    let saved_digit = cursor.get_ref()[saved_pos];
+                    if let Some(precision) = precision {
+                        try!(write!(cursor, "{:.*}s", cmp::min(precision, 9), n));
+                    } else {
+                        // variable precision up to 3 decimals
+                        try!(write!(cursor, "{:.3}", n));
+                        backup_over_zeros(&mut cursor);
+                        try!(write!(cursor, "s"));
+                    }
+                    // make sure we didn't round up to 1.0 when printing the float
+                    if cursor.get_ref()[saved_pos] == b'1' {
+                        // we did. back up and rewrite the seconds
+                        if s == u64::max_value() {
+                            // we can't represent a larger value. Luckily, we know that
+                            // u64::max_value() ends with '5', so we can just replace it with '6'.
+                            cursor.get_mut()[saved_pos] = b'6';
+                        } else {
+                            try!(cursor.seek(SeekFrom::Start(0)));
+                            try!(write!(cursor, "{}", s+1));
+                            match precision {
+                                None | Some(0) => {},
+                                Some(precision) => {
+                                    // we need to write out the trailing zeroes
+                                    try!(write!(cursor, "."));
+                                    for _ in 0..cmp::min(precision, 9) {
+                                        try!(write!(cursor, "0"));
+                                    }
+                                }
+                            }
+                            try!(write!(cursor, "s"));
+                        }
+                    } else {
+                        // restore the digit we overwrite earlier
+                        cursor.get_mut()[saved_pos] = saved_digit;
+                    }
+                }
+            }
+            // buf now contains our fully-formatted value
+            let pos = cursor.position() as usize;
+            let buf = &cursor.into_inner()[..pos];
+            // formatting always writes strings
+            Ok(unsafe { str::from_utf8_unchecked(buf) })
         }
+
+        let result = write_to_buf(self, f, &mut buf);
+        // writing to the stack buffer should never fail
+        debug_assert!(result.is_ok(), "Error in <Duration as Display>::fmt: {:?}", result);
+        let buf_str = try!(result.or(Err(fmt::Error)));
+        // If fmt::Formatter::with_padding() was public, this is where we'd use it.
+        write!(f, "{}", buf_str)
```

Comment:
We actually can't use `f.pad(buf_str)`, because it interprets precision as the maximum width of the string to print, but we're using precision here for a different purpose. I'm actually kind of surprised the formatting code doesn't have any way of changing the flags, and I decided not to attempt to hack it with a `write!()` call that uses a caller-supplied width because that discards alignment and fill.